### ISSUE TITLE
Real time dashboards for Beam

### ISSUE THREAD

**Issue by REASY** [LBNL-UCB-STI/beam#2272]
### Why we need it? 
We have a way to see what happens with simulation, but it is not in real time. We have to wait when the iteration is over and then we can explore the output folder of simulation which contains PNGs and CSVs files. This way is is fine for experienced Beam users, but having real time dashboards for Beam will make it:

- More user-friendly
- Allow to see the progress of simulation right now
- Allow to change the way how the data is represented without the need of code changes (this statement is true only for the case when the same metric is represented differently)

### Proposed architecture
I'm proposing to use the following well-known technologies to achieve it:
- [Grafana](https://grafana.com/) for time series analytics
- [InfluxDB Community](https://community.influxdata.com/) as time series database
- [Docker](https://www.docker.com/) as containerization. It will allow us to prepare our own environment which is easy to run and deploy

![architecture](https://user-images.githubusercontent.com/5107562/68531985-f15ba600-034a-11ea-9fac-49de2b997427.png)

### Technical details
I'd suggest to go with single docker image which contains Grafana + InfluxDB. I've played with https://hub.docker.com/r/philhawthorne/docker-influxdb-grafana, so far so good. We have to make sure that the log folders and data folders are mounted to host machine so we can copy it at the end of simulation to the output folder of beam and copy the results over S3 (as we are doing currently). I've already done some work related to this issue. It's in https://github.com/LBNL-UCB-STI/beam/pull/2269

[Very important note regarding InfluxDB](https://github.com/influxdata/influxdb/issues/2055):
>>> Points in a series can not have the same exact time (down to nanosecond). A series is defined by the measurement and tagset. 

Why it is important ? Because there are cases when different people have the same end activity time. In the following screenshot you can see that on time `30240` there were 518 `ModeChoice` events with `car` mode:
![image](https://user-images.githubusercontent.com/5107562/68532068-ed7c5380-034b-11ea-8fb5-a3cd3a5bd5b1.png)
As a workaround we could keep track metric name and its timestamp and if it is the same, just add extra 1 microsecond to it:
```scala
import java.time.{LocalDate, ZoneId}
import java.util.concurrent.{ConcurrentHashMap, TimeUnit}

object Metrics {
  val metricToLastSeenTs: ConcurrentHashMap[String, Long] = new ConcurrentHashMap[String, Long]()
  val step: Long = TimeUnit.MICROSECONDS.toNanos(1L)

  def influxTime(metricName: String, simulationTimeSeconds: Int): Long = {
    val todayBeginningOfDay = LocalDate.now().atStartOfDay()
    val todayInstant = todayBeginningOfDay.toInstant(ZoneId.systemDefault().getRules.getOffset(todayBeginningOfDay))
    val ts = TimeUnit.MILLISECONDS.toNanos(todayInstant.toEpochMilli) + TimeUnit.SECONDS.toNanos(simulationTimeSeconds)
    Metrics.getNextInfluxTime(metricName, ts)
  }

  def getNextInfluxTime(metricName: String, tsMicroSeconds: Long): Long = {
    // See https://github.com/influxdata/influxdb/issues/2055
    // Points in a series can not have the same exact time (down to nanosecond). A series is defined by the measurement and tagset.
    // We store the last seen `tsMicroSeconds` and add up `step` in case if it is already there
    val key = s"${metricName}:$tsMicroSeconds"
    val prevTs = metricToLastSeenTs.getOrDefault(key, tsMicroSeconds)
    val newTs =  prevTs + step
    metricToLastSeenTs.put(key, newTs)
    newTs
  }
}
```
This is not perfect solution due to the thing that now every pair (metric, timestamp) should be stored in `ConcurrentHashMap`. So it is extra overhead by CPU and memory.

Use [influxdb-java](https://github.com/influxdata/influxdb-java) as JVM client to InfluxDB 1.x

### List of faced issues

- When you want to automatically upload your Grafana dashboard, make sure you get its JSON representation via API call, not via exporting it via UI (https://github.com/grafana/grafana/issues/11356)
- When you try to upload previously exported dashboard JSON via API call `POST /api/dashboards/db `, it can get the error `404 Dashboard not found` .  You should remove `dashboard.id` field: https://github.com/grafana/grafana/issues/7584

**Comment by irishwithaxe** [comment]
need to implement way to write global (not per iteration) metrics

**Comment by irishwithaxe** [comment]
need to implement way to fix time series timestamp issue (when we cant write two points with same timestamp separately)

**Comment by irishwithaxe** [comment]
we expect to not have more than 1000 metrics at all

**Comment by nikolayilyin** [comment]
mostly implemented in PR #2269 
the rest will be implemented in issues #2435, #2418
